### ISSUE TITLE
Client nodes not working with table elements

### ISSUE THREAD

**Issue by paurkedal** [ocsigen/eliom#307]
With Eliom 5.0.0 the following program attaches the client-side `<td/>` and `<tr/>` elements to the top level instead of inside the table.  In particular the "one" and "three" is correctly placed, while "two" and "second line" are detached.
```ocaml
[%%shared
  open Eliom_content.Html5
]

let handler () () = Lwt.return @@
  Eliom_tools.D.html ~title:"Test <td/> Nodes"
    (D.body [
      D.table ~a:[F.a_style "border: solid thin green"] [
        D.tr [
          D.td [D.pcdata "one"];
          C.node [%client D.td [D.pcdata "two"]];
          D.td [C.node [%client D.pcdata "three"]];
        ];
        C.node [%client D.tr [D.td [D.pcdata "second line"]]];
      ]
    ])

module App = Eliom_registration.App
  (struct let application_name = "test_td_nodes" end)
let service =
  App.register_service ~path:[] ~get_params:Eliom_parameter.unit handler
```

**Comment by paurkedal** [comment]
The issue is also reproduced in the development version with the modified code:
```ocaml
[%%shared
  open Eliom_content.Html
]

let handler () () = Lwt.return @@
  Eliom_tools.D.html ~title:"Test <td/> Nodes"
    (D.body [
      D.table ~a:[F.a_style "border: solid thin green"] [
        D.tr [
          D.td [D.pcdata "one"];
          C.node [%client D.td [D.pcdata "two"]];
          D.td [C.node [%client D.pcdata "three"]];
        ];
        C.node [%client D.tr [D.td [D.pcdata "second line"]]];
      ]
    ])

module App = Eliom_registration.App
  (struct
    let application_name = "test_td_nodes"
    let global_data_path = None
  end)
let service =
  App.create ~id:Eliom_service.(Path [])
             ~meth:Eliom_service.(Get Eliom_parameter.unit)
             handler
```

**Comment by vasilisp** [comment]
The HTML we send contains a `span` element as a placeholder that is later replaced by the client node. Apparently browsers don't like `span` inside tables. A work-around is to provide an `~init` value:

```ocaml
let handler () () = Lwt.return @@
  Eliom_tools.D.html ~title:"Test <td/> Nodes"
    (D.body [
        D.table ~a:[F.a_style "border: solid thin green"] [
          D.tr [
            D.td [D.pcdata "one"];
            C.node ~init:(D.td [])
              [%client D.td [D.pcdata "two"]];
            D.td [C.node [%client D.pcdata "three"]];
          ];
          C.node ~init:(D.tr [])
            [%client D.tr [D.td [D.pcdata "second line"]]];
        ];
      ])
```

Automatically building a correct `init` for the polymorphic `C.node` function sounds hard, given that we can't really see what kind of element the client code produces. But we could in principle provide `C.div`, `C.tr` etc.

**Comment by vasilisp** [comment]
At the very least, we need to document the limitations.

**Comment by paurkedal** [comment]
I see the problem.  I think this is what [the `<template/>` element](https://www.w3.org/TR/html5/scripting-1.html#the-template-element) is for.  I have never used it, just found it now after reading your first comment.

The workaround is quite reasonable, esp. if it's documented.  About the other route, I assume that would read `C.td [%client D.td [...]]`, so that the edge element is repeated on the client side?  Or the `C.td` element would itself be fix, but allow client-side content?

**Comment by vasilisp** [comment]
I was not aware of `<template>`, thanks. It could be useful in a few other places in Eliom.

For `C.td`, we could implement either option, but either would require reworking the TyXML functorial interface (CC @Drup). Whether this is worth the trouble, I don't know.

Updated documentation: [Eliom_content.Html.C](https://ocsigen.org/eliom/dev/api/server/Eliom_content.Html.C).

**Comment by Drup** [comment]
I don't think the template element is very useful to us (it's basically a piece of html that you save in the page but is not showed, not a placeholder element that can be replaced later). In particular, template in table is no more valid than span in table.

About the other question ... what would be semantics of attributes ? What do we do with elements that can't be empty ? It sounds like we would quickly come up with something that is equivalent to the shared react stuff.
I kinda like the fact that we have only two functions there.

**Comment by paurkedal** [comment]
@Drup: That seems to be the primary purpose, but the content model of `<table/>` and `<tr/>` lists script-supporting elements, which include `<template/>`. http://www.html5rocks.com/en/tutorials/webcomponents/template/ "The Pillars" point 4 seems to back that interpretation.

**Comment by Drup** [comment]
Hum, indeed, I looked at [mozilla's documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template) quickly, where it was not specified clearly.

So, that would made a suitable default element, but according to [caniuse](http://caniuse.com/#search=template), it's probably too bleeding edge. Maybe we can just document it.

**Comment by paurkedal** [comment]
Yes, at least on my workplace the lack of IE support would be a problem.  But if `<template/>` is added, would it not be natural to call it `C.template`, which could come with a warning about browser support, pointing most people to `C.node` for now?

I agree that it's better not to fill up the `C` module with the whole HTML5 zoo.  A few shortcuts defined in terms of `C.node ~init:...` where span is invalid might suffice, though given the prospect of `<template/>`, using `~init` for now seems perfectly fine too.

**Comment by Drup** [comment]
I agree with both your "Update". I think we just need to improve the documentation a bit.
